### PR TITLE
ebooks: create holdings automatically after record harvesting

### PR DESCRIPTION
**Pull request by BadrAly**
* Adds online type to item types resource.
* Adds is_online to location resource.
* Adds online_harvested_source to the organisation resource.
* Adds extended validation for the resource location field is_online.
* Add extended validation for the resource item_type field type.
* Adds an option to setup script to continue process after holding_items generation.
* Fixes some variables snake cases.
* Fixes typos.

Co-Authored-by: Alicia Zangger <alicia.zangger@rero.ch>
Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>
Co-Authored-by: Aly Badr <aly.badr@rero.ch>

### CODE REVIEW

File: rero_ils/modules/api.py [hunk 1]
```diff
@@ -101,6 +103,20 @@ class IlsRecord(Record):
     object_type = 'rec'
     indexer = IlsRecordIndexer
 
+    def validate(self, **kwargs):
+        """Validate record according to schema and specific rules."""
```
Comment (iGormilhit) [inline rero_ils/modules/api.py]: Could `specific rules` be… more specific? Or is it documented elsewhere?

File: rero_ils/modules/ebooks/utils.py [hunk 2]
```diff
@@ -49,3 +55,67 @@ def add_oai_source(name, baseurl, metadataprefix='marc21',
             db.session.commit()
             return 'Updated'
         return 'Not Updated'
+
+
+def create_document_holding(record):
+    """Create a document and a holding for an ebook."""
```
Comment (iGormilhit) [inline rero_ils/modules/ebooks/utils.py]: Is it only for an harvested ebook? If yes, then be more specific.

File: rero_ils/modules/ebooks/utils.py [hunk 3]
```diff
@@ -49,3 +55,67 @@ def add_oai_source(name, baseurl, metadataprefix='marc21',
...
+                        create_holding(
+                            document_pid=new_record.pid,
+                            location_pid=location,
+                            item_type_pid=item_type_pid,
+                            electronic_location=harvested_source)
+    return new_record
+
+
+def update_document_holding(record, pid):
+    """Update a document and a holding for an ebook."""
```
Comment (iGormilhit) [inline rero_ils/modules/ebooks/utils.py]: Same as previous comment.

File: rero_ils/modules/errors.py [hunk 4]
```diff
@@ -32,3 +32,11 @@ class PolicyNameAlreadyExists(RecordsError):
 
 class InvalidRecordID(RecordsError):
     """Error raised when the ID of record in invalid."""
```
Comment (iGormilhit) [inline rero_ils/modules/errors.py]: is instead of in?

File: rero_ils/modules/item_types/api.py [hunk 5]
```diff
@@ -59,6 +60,20 @@ class ItemType(IlsRecord):
     fetcher = item_type_id_fetcher
     provider = ItemTypeProvider
 
+    def extended_validation(self, **kwargs):
+        """Returns reasons for validation failures, otherwise True."""
```
Comment (iGormilhit) [inline rero_ils/modules/item_types/api.py]: Should what's being validated here be explicit in the docstring?

File: rero_ils/modules/locations/api.py [hunk 6]
```diff
@@ -55,6 +57,14 @@ class Location(IlsRecord):
     fetcher = location_id_fetcher
     provider = LocationProvider
 
+    def extended_validation(self, **kwargs):
+        """Returns reasons for validation failures, otherwise True."""
```
Comment (iGormilhit) [inline rero_ils/modules/locations/api.py]: Same as precedent comment.

File: rero_ils/modules/holdings/api.py [hunk 7]
```diff
@@ -171,11 +172,17 @@ def get_items_filter_by_viewcode(self, viewcode):
...
-                if (item.organisation_pid == org_pid):
+                if item.organisation_pid == org_pid:
                     items.append(item)
             return items
         return holdingItems
 
+    @property
+    def get_items(self):
+        """Return items of holding record."""
+        return [Item.get_record_by_pid(item_pid)
```
Comment (rerowep) [inline rero_ils/modules/holdings/api.py]: Are there a lot of items? Should it be better to use `yield`

File: rero_ils/modules/libraries/api.py [hunk 8]
```diff
@@ -67,6 +67,29 @@ class Library(IlsRecord):
...
+            return next(result).pid
+        except StopIteration:
+            return None
+
+    def get_online_locations(self):
+        """Return online locations."""
+        search = LocationsSearch()\
+            .filter('term', is_online=True)\
+            .filter('term', library__pid=self.pid)
+        result = search.source(['pid']).scan()
```
Comment (rerowep) [inline rero_ils/modules/libraries/api.py]: Not very different `online_location` and `get_online_locations`

**Pull request merged**